pkg/cluster/internal/providers/docker/provision.go

```diff
@@ -320,6 +328,79 @@ func getSubnets(networkName string) ([]string, error) {
 	return strings.Split(strings.TrimSpace(lines[0]), " "), nil
 }
 
+func getUsedIPs(networkName string) ([]string, error) {
+	format := `{{range (index . "Containers")}}{{index . "IPv4Address"}} {{end}}`
+	cmd := exec.Command("docker", "network", "inspect", "-f", format, networkName)
+	lines, err := exec.OutputLines(cmd)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get used ip addresses")
+	}
+
+	usedIPs := strings.Split(strings.TrimSpace(lines[0]), " ")
+	for i := range usedIPs { // Trim the CIDR notation
+		usedIPs[i] = strings.Split(usedIPs[i], "/")[0]
+	}
+
+	return usedIPs, nil
+}
+
+func generateNodeLabels(node *config.Node) []string {
+	args := []string{}
+	for k, v := range node.ExtraLabels {
+		args = append(args, "--label", fmt.Sprintf("%s=%s", k, v))
+	}
+
+	return args
+}
+
+// generateIPForNode generates a static IP that can be used for this node
+// on the docker network
+func generateIPForNode(networkName string) ([]string, error) {
+	subnets, err := getSubnets(networkName)
+	if err != nil {
+		return nil, err
+	}
+
+	var ipv4Subnet *net.IPNet
+	for _, subnet := range subnets {
+		ip, cidr, err := net.ParseCIDR(subnet)
+		if err != nil {
+			continue
+		}
+
+		// Skip ipv6 addresses (for now)
+		if ip.To4() == nil {
+			continue
+		}
+
+		// grab the first ipv4 cidr we find
+		ipv4Subnet = cidr
+		break
+	}
+
+	usedIPs, err := getUsedIPs(networkName)
+	if err != nil {
+		return nil, err
+	}
+
+	ipam := goipam.New()
+	ipam.NewPrefix(ipv4Subnet.String())
+
+	// First IP in the block is always being used, so skip it.
+	ipam.AcquireIP(ipv4Subnet.String())
+
+	for _, ip := range usedIPs {
+		ipam.AcquireSpecificIP(ipv4Subnet.String(), ip)
```

Comment:
Nit: Can this cause trouble if the `ip` is from a different subnet? Asking cause we don't get used IPs by subnet.

Reply:
If you don't even get the error, go just proceeds without complaining?

Reply:
It could, but we're reading the IPs from `docker network inspect` so we can reasonably assume they are in the subnet, even if they aren't it'd fail (and return an err, which we ignore here) and noop (which is fine, we only care about ones in our subnet). go ipam returns an error if we didn't `AcquirePrefix` a subnet before hand, so they would error if outside of the subnet.

Reply:
This is really just nitpicking for learning purposes. Theoretically you can have multiple ipv4 subnets on docker network. I'm pretty sure we would not do that.